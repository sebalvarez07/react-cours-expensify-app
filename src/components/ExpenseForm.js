import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

class ExpenseForm extends React.Component {
    constructor(props){
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            amount: props.expense ? (props.expense.amount / 100).toString(): '',
            note: props.expense ? props.expense.note : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: undefined
        };
    }

    onDescriptionChange = (e) => {

        /* 
            We do this as a two step process because setState is an async function meaning that the object contents do not get executed right away
            This is a problem in this case as the value of e.target does not 'persist', and must be used right away.
            There is another way to go about this if we want to use description: e.target.value inside the setState
            By putting this line before everything :
            e.persist()
        */
       
        const description = e.target.value;
        this.setState(() => ({ description }));
    }

    onAmountChange = (e) => {
        const amount = e.target.value;

        // The $ sign at the end is crucial as it marks the end of the input. Anything written after the 
        // 2 decimal places will not match and so it will not be written on the state, which means it will
        // Not be printed as a value to the input
        // Thus the input is registered onChange, but not displayed, the process of dealing with the input
        // Handled and the when the state is updated, the value={this.state.amount} will be printed

        /*
            ^   - marks the begining of input
            \d  - Means digits = [0 - 9]
            *   - Means match the preceeding expression 0 or more times = {0,}
            ()? - Means match preceeding expression 0 or 1 time (maybe it happens, maybe not) = {0, 1}
            \.  - Means literally a dot(.)
            $   - Marks the end of input. IF this isn't the end, returns false

            !amount --> If there is no amount OR if there is a match
        */

        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(() => ({ amount }));
        }
    }

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    }

    onDateChange = (createdAt) => {
        
        if(createdAt){
            this.setState(() => ({ createdAt: createdAt }));
        }
    }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    }

    handleSubmit = (e) => {
        e.preventDefault();

        if(!this.state.description || !this.state.amount){
            this.setState(() => ({error: 'Please provide a description and amount'}));
        } else {
            this.setState(() => ({error: '' }));
            this.props.onSubmit({
                description: this.state.description, 
                amount: parseFloat(this.state.amount, 10) * 100,    // We eliminate the . of decimals while keeping the decimes in the ciphre
                note: this.state.note,
                createdAt: this.state.createdAt.valueOf()
            });
        }
    }

    render() {
        return (
            <form onSubmit={this.handleSubmit}>

                {this.state.error && <h3>{this.state.error}</h3>}

                <label>Description</label>
                <input 
                    type="text" 
                    name="description" 
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                />
        
                <label>Amount</label>
                <input 
                    type="text" 
                    name="amount"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                />
        
                <label>Note</label>
                <textarea 
                    name="note"
                    placeholder="Add a note for your expense (optional)"
                    value={this.state.note}
                    onChange={this.onNoteChange}
                />

                <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange} 
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
        
                <button>Add Expense</button>
            </form>
        )
    }
}

export default ExpenseForm;