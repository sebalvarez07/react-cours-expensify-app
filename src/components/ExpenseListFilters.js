import React from 'react';
import { connect } from 'react-redux';
import { setFilterText, sortByDate, sortByAmount, setStartDate, setEndDate} from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {

    state = {
        calendarFocused: null
    };

    handleOnDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocuseChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }))
    };

    onChangeTextFilter = (e) => {
        const val = e.target.value;
        this.props.setFilterText(val);
    };

    onChangeSortbyFilter = (e) => {
        const val = e.target.value;

        if(val === 'date') {
            this.props.sortByDate();
        } 
        else if(val === 'amount') {
            this.props.sortByAmount();
        }
    };

    render(){
        return (
            <div>
                <input 
                    type="text"  
                    value={this.props.filters.text} 
                    onChange={this.onChangeTextFilter}
                />
                <select 
                    value={this.props.filters.sortBy} 
                    onChange={this.onChangeSortbyFilter}>
                    
                    <option value="amount">Amount</option>
                    <option value="date">Date</option>
                </select>
        
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate} 
                    onDatesChange={this.handleOnDatesChange}
                    focusedInput={this.state.calendarFocused} 
                    onFocusChange={this.onFocuseChange}
                    isOutsideRange={() => false}
                    numberOfMonths={1}
                    showClearDates={true}
                />
            </div>
        );
    }
};

const mapDispatchToProps = (dispatch) => {
    return {
        setStartDate: (startDate) => dispatch(setStartDate(startDate)),
        setEndDate: (endDate) => dispatch(setEndDate(endDate)),
        setFilterText: (text) => dispatch(setFilterText(text)),
        sortByDate: () => dispatch(sortByDate()),
        sortByAmount: () => dispatch(sortByAmount()),
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}

const ExpenseListFiltersWrapped = connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);

export default ExpenseListFiltersWrapped;