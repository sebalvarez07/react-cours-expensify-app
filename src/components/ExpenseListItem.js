import React from 'react';
import { Link } from 'react-router-dom'

// We can destructure as we're passing the object expense's contents as a set of props (using spread operator)
const ExpenseListItem = ({id, description, amount, createdAt}) => (
    <div>
        <h3>
            <Link to={`/edit/${id}`} >Description: {description}</Link>
        </h3>
        
        <p>Amount: {amount}</p>
        <p>createdAt {createdAt}</p>
        <br></br> 
    </div>
);

export default ExpenseListItem;