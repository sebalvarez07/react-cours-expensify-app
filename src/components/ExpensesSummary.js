import React from 'react';
import { connect } from 'react-redux';
import expensesTotal from '../selectors/expenses-total';
import getVisibleExpenes from '../selectors/expenses';
import numeral from 'numeral';

export const ExpensesSummary = ({expenseCount, expensesTotal}) => {

    const word = expenseCount === 1 ? 'expense' : 'expenses';
    const numFormat = numeral(expensesTotal / 100).format('$0,0.00');
    return (
        <div>
            <h2>Expenses summary: Viewing {expenseCount} {word} totaling {numFormat}</h2>
        </div>
    )
};

const mapStateToProps = (state) => { 
    return {
        expenseCount: getVisibleExpenes(state.expenses, state.filters).length,
        expensesTotal: expensesTotal(getVisibleExpenes(state.expenses, state.filters))
    }
};

export default connect(mapStateToProps)(ExpensesSummary);
