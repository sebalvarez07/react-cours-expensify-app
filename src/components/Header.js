import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLoginOut } from '../actions/auth';

export const Header = ({ startLoginOut }) => (
    <header className="header">
        <div className="content-container">
            <div className="header__content">
                <Link className="header__title" to="/dashboard" >
                    <h1>Expensify</h1>
                </Link>
                <button 
                    className="btn btn--link"
                    onClick={startLoginOut}
                    >
                    Logout
                </button>
            </div>
        </div>
        
    </header>
);

const mapDispatchToProps = (dispatch) => {
    return {
        startLoginOut: () => dispatch(startLoginOut())
    }
}
 
export default connect(undefined, mapDispatchToProps)(Header);