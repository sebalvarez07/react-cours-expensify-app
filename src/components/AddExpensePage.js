import React from 'react';
import ExpenseForm from './ExpenseForm';
import  { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {

    onSubmit = (expense) => {
        this.props.startAddExpense(expense);
        {/** 
            We use redirect, as we do not have a Link that we can simply click to send us to a new page
            So we use history.push (which we get as a prop for being a Router component - see AppRouter.js under the routers folder)
            This redirects us to any of the paths that are Router components

            NOTE: Using push also allow us to go back through the browsers back button as if it were actually a multi-page site
        */}
        this.props.history.push('/');           
    }

    render() {
        return (
            <div>  
                <h2>This is from my add Expense component</h2>
                <ExpenseForm onSubmit={this.onSubmit}/>
            </div>
        )
    }
}

// Function gives us access to dispatch, and allows us to return an object that turns its properties into props properties
const mapDispatchToProps = (dispatch) => ({ startAddExpense: (expense) => dispatch(startAddExpense(expense)) });

export default connect(undefined, mapDispatchToProps)(AddExpensePage);