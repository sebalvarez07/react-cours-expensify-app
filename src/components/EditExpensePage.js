import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import ReactModal from 'react-modal';
 
ReactModal.setAppElement(document.getElementById('app'));

export class EditExpensePage extends React.Component {

    state = {
        showModal: false
    }
    
    onSubmit = (updates) => {
        this.props.startEditExpense(this.props.expense.id, updates);
        this.props.history.push('/');
    };

    handleRemoveExpense = () => {
        this.props.startRemoveExpense({id: this.props.expense.id});
        this.setState({ showModal: false });
        this.props.history.push('/');
    };

    handleOpenModal = () => {
        this.setState({ showModal: true });
    }
      
    handleCloseModal = () => {
        this.setState({ showModal: false });
    }
      

    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h2 className="page-header__title">
                            Edit Expense
                        </h2>
                    </div> 
                </div>

                <div className="content-container">
                    <ExpenseForm 
                        expense={this.props.expense}
                        onSubmit={this.onSubmit}
                    />
                    <button className="btn btn--grey" onClick={this.handleOpenModal}>
                        Remove Expense
                    </button>
                </div>

                <ReactModal 
                    className='modal modal--remove-expense'
                    overlayClassName='modal-overlay'
                    isOpen={this.state.showModal}>
                    <h3 className="modal__title">Delete Expense</h3>
                    <p className="modal__text">Are you sure you want to delete this expense?</p>
                    <div className="modal__footer">
                        <button className="btn btn--grey" onClick={this.handleCloseModal}>
                            Close
                        </button>

                        <button className="btn btn--blue remove-expense__btn" onClick={this.handleRemoveExpense}>
                            Remove expense
                        </button>
                    </div>
                </ReactModal>
            </div>
        )
    };
}

const mapDispatchToProps = (dispatch) => {
    return {
        startEditExpense: (id, updates) => dispatch(startEditExpense(id, updates)),
        startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
    }
}

/*
    state gives the store's state
    props gives us the connected (or wrapped) props
*/
const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find(el => el.id === props.match.params.id)
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);