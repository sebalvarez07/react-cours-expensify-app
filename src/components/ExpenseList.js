import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getVisibleExpenes from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-desktop">Expense</div>
            <div className="show-for-desktop">Amount</div>
        </div>
        <div className="list-body">
            {
                props.expenses.length === 0 ? (
                    <div className="list-item list-item--message">
                        <span>No expenses</span>
                    </div>
                ) : (
                    props.expenses.map(
                        (expense, index) => <ExpenseListItem key={`${expense.id}`} {...expense}/>
                    )
                )
            }
        </div>
        
    </div> 
);

/*
    - This is called everytime the store changes
    - Returns an object with the data that this component needs and sends it to the component as props
    - Returns an object with expenses and dispatch fn
*/
const mapStateToProps = (state) => {
    return {
        expenses: getVisibleExpenes(state.expenses, state.filters)
    };
};
/*
    - connect() returns a new function that takes the component to be wrapped as an argument
    - We don't have to be concerned with store.subscribe() or dispatching. All taken care of by react-redux
*/
export default connect(mapStateToProps)(ExpenseList);