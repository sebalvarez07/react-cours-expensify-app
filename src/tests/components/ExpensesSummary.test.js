import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('Should render summary with 2 expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={3} expensesTotal={200} />);
    expect(wrapper).toMatchSnapshot();
});

test('Should render summary with 1 expense', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={100} />);
    expect(wrapper).toMatchSnapshot();
});