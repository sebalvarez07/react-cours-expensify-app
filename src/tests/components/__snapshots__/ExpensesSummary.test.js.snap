// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render summary with 1 expense 1`] = `
<div>
  <h2>
    Expenses summary: Viewing 
    1
     
    expense
     totaling 
    $1.00
  </h2>
</div>
`;

exports[`Should render summary with 2 expenses 1`] = `
<div>
  <h2>
    Expenses summary: Viewing 
    3
     
    expenses
     totaling 
    $2.00
  </h2>
</div>
`;
